requre/objects.py

```diff
@@ -149,30 +151,51 @@ def internal(*args, **kwargs):
         return internal
 
     @classmethod
-    def decorator(cls, *, item_list: list) -> Any:
+    def decorator(cls, *, item_list: list, apply_arg_filters: Dict = {}) -> Any:
         """
         Class method for what should be used as decorator of import replacing system
         This use list of selection of *args or **kwargs as arguments of function as keys
 
         :param item_list: list of values of *args nums,  **kwargs names to use as keys
+        :param apply_arg_filters: dict of function to apply to keys before storing
+                                  (have to be listed in item_list)
         :return: output of func
         """
-
         def internal(func: Callable):
             @functools.wraps(func)
             def internal_internal(*args, **kwargs):
                 keys = cls.get_base_keys(func)
-                for item in item_list:
-                    if isinstance(item, int):
-                        keys.append(args[item])
+                # get all possible arguments of passed function
+                arg_keys = inspect.getfullargspec(func)[0]
+                for param_name in item_list:
+                    # if you pass int as an agrument, it forces to use args
+                    if isinstance(param_name, int):
+                        key = args[param_name]
+                    # try to look into kwargs if item is there
+                    elif param_name in kwargs:
+                        key = kwargs[param_name]
+                    #  translate param name to positional argument
+                    else:
+                        # out of index check. This is bad but possible use case, raise warning and continue
+                        if len(args) <= arg_keys.index(param_name):
+                            warnings.warn(f"You've defined keys: {item_list} but '{param_name}' is not part"
+                                          f" of args:{args} and kwargs:{kwargs},"
+                                          f" original function and args: {func.__name__}({arg_keys})")
+                            # but add there None as key, to not spoil dictionary deep
+                            key = None
+                        else:
+                            key = args[arg_keys.index(param_name)]
+                    if param_name not in apply_arg_filters:
+                        keys.append(key)
                     else:
-                        keys.append(kwargs[item])
+                        keys.append(apply_arg_filters[param_name](key))
```

Comment:
We can have mapping/replacing and filtering on multiple places:
- on keys before the original function execution (some injection, preprocessing) (map only)
- on keys after the run but will be used for matching keys on a future run (to allow custom matchers) (map only)
- on values after execution to allow some postprocessing (to allow removing sensitive data from the response) (map or filter)

Reply:
> We can have mapping/replacing and filtering on multiple places:
> 
>     * on keys before the original function execution (some injection, preprocessing) (map only)

It sounds interesting, but not sure if it make any sense, because it means change value for function execution,
but this could be used for fuzzing, but we are "mocking" on the 2nd or 3th level, so it will test the libraries, not executed code.
But may be useful also to change returned values,somethinglike `ogr.github.getproject("ogr")` and you decide to use helloworld project and do not want to change original code.
> 
>     * on keys after the run but will be used for matching keys on a future run (to allow custom matchers) (map only)
> 
It is this case (especially remove unwanted or sensitive data from keys). 
Idea with filtering keys seems interesting but it may lead to unpleasant changes what I've already when playing with custom deep of dict and variable keys and it is that you cannot have dict and list on some level, and one value will expect that something is leaf, but it will provide also deeper structures. 

>     * on values after execution to allow some postprocessing (to allow removing sensitive data from the response) (map or filter)

We have it already implemented as offline postprocessing filters, I understand that it make sense to have it also part of this, so that it will be easier to read, what filter which file. as an cond, it will be harder to use for some generic matches. (e.g.: remove timestamps. you have to have it within each testcase, so it does not make sense)
But both ways could live together, and some specific items remove as part of exact case and some remove as files postprocessing.